ast/edit/match.go

```diff
@@ -0,0 +1,483 @@
+package edit
+
+import (
+	"fmt"
+
+	"github.com/influxdata/flux/ast"
+)
+
+// Match takes an AST and a pattern and returns the nodes of the AST that
+// match the given pattern.
+// The pattern is an AST in turn, but it is partially specified, i.e. nil nodes are ignored
+// while matching.
+// In the case of slices (e.g. ast.File.Body, ast.CallExpression.Arguments) the matching
+// mode can be "exact" or "fuzzy".
+// Let ps be a slice of nodes in the pattern and ns in node.
+// In "exact" mode, slices match iff len(ps) == len(ns) and every non-nil node ps[i] matches ns[i].
+// In "fuzzy" mode, slices match iff ns is a superset of ps; e.g. if ps is empty, it matches every slice.
+func Match(node ast.Node, pattern ast.Node, matchSlicesFuzzy bool) []ast.Node {
+	var sms sliceMatchingStrategy
+	if matchSlicesFuzzy {
+		sms = &fuzzyMatchingStrategy{}
+	} else {
+		sms = &exactMatchingStrategy{}
+	}
+	mv := &matchVisitor{
+		matched: make([]ast.Node, 0),
+		pattern: pattern,
+		sms:     sms,
+	}
+	ast.Walk(mv, node)
+	return mv.matched
+}
+
+type matchVisitor struct {
+	matched []ast.Node
+	pattern ast.Node
+	sms     sliceMatchingStrategy
+}
+
+func (mv *matchVisitor) Visit(node ast.Node) ast.Visitor {
+	if match(mv.pattern, node, mv.sms) {
+		mv.matched = append(mv.matched, node)
+	}
+	return mv
+}
+
+func (mv *matchVisitor) Done(node ast.Node) {}
+
+type sliceMatchingStrategy interface {
+	matchFiles(p, n []*ast.File) bool
+	matchImportDeclarations(p, n []*ast.ImportDeclaration) bool
+	matchStatements(p, n []ast.Statement) bool
+	matchProperties(p, n []*ast.Property) bool
+	matchExpressions(p, n []ast.Expression) bool
+}
+
+type fuzzyMatchingStrategy struct{}
```

Comment:
For slice matching strategies I merely copy pasted the code. I don't want to copy the slice to get a `[]ast.Node`. I thought about code generation, but I didn't want to lose too much time on it. Is this the way to go?

Reply:
I think this is fine for now.  You have what you need for the test refactor tool.

Moving forward, maybe there will be a `Matcher` object, and you can have an `Options` class that uses varargs similar to how the Planner handles options:
https://github.com/influxdata/flux/blob/master/plan/logical.go#L24
This way the API doesn't need to change when new options are added (for example, if there is an additional predicate that must pass for matches).